[SPARK-53399][SQL] Restore CollapseProject to merge unrelated expressions

peter-toth opened the pull request:
### What changes were proposed in this pull request?

Latest improvements to `CollapseProject` rule (like https://github.com/apache/spark/pull/33958) prevented duplicating expressive expressons, which brings considerable performance improvement in many cases.
But there is one particular case when it can introduce significant perfoamance degradation. Consider a query where the adjacent project nodes don't get collapsed because they contain expensive, multiple referenced expressions, but the nodes also contain Python UDF expressions that otherwise wouldn't prevent project node collapsion. E.g.:
```
Project a + a as a_plus_a, PythonUDF(...) as udf2, udf1
  Project <expensive calculation> as a, PythonUDF(...) as udf1
    ...
```
In the above example `CollapseProject` doesn't modify the 2 project nodes, which then causes 2 `BatchEvalPython` nodes to appear in the plan when `ExtractPythonUDFs` extracts them:
```
Project a + a as a_plus_a, udf2, udf1
  BatchEvalPython PythonUDF(...) -> udf2
    Project <expensive calculation> as a, udf1
      BatchEvalPython PythonUDF(...) -> udf1
        ...
```
The 2 `BatchEvalPython` nodes can cause significant serialization/deserialization overhead compared to the case when the original project nodes were collapsed and we had only 1 `BatchEvalPython` node.

The old behaviour can be restored with setting `spark.sql.optimizer.collapseProjectAlwaysInline=true`, but it is still not ideal as we lose the performance improvement in other cases.

This PR improves to the `CollapseProject` rule to not just collapse or don't collapse adjacent nodes, but be able to decide on individual expressions if it makes sense to merge them with expressions from the other node.

After this PR the rule modifies the examle plan and merges the unrelated `PythonUDF`s into the consumer `Project` node, while it keeps the related `a` and `a + a as a_plus_a` expressions separate:
```
Project a + a as a_plus_a, PythonUDF(...) as udf2, PythonUDF(...) as udf1
  Project <expensive calculation> as a
    ...
```
The plan is then transformed to:
```
Project a + a as a_plus_a, udf2, udf1
  BatchEvalPython PythonUDF(...) -> udf2, PythonUDF(...) -> udf1
    Project <expensive calculation> as a
      ...
```
and both Python UDFs are calculated in one run.

### Why are the changes needed?

To fix performance regression caused by latest changes to `CollapseProject`.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

New and existing UTs.

### Was this patch authored or co-authored using generative AI tooling?

No.

peter-toth committed:
[SPARK-53399][SQL] Improve CollapseProject

dongjoon-hyun reviewed (commented):
Since this is a regression bug fix, we had better be clear in the PR title instead of `Improve CollapseProject`. Also, please revise the `Affected Versions` if you want to backport this to live release branches, @peter-toth .

<img width="495" height="283" alt="Screenshot 2025-08-27 at 13 35 51" src="https://github.com/user-attachments/assets/9b289d2d-e26d-44ed-86ec-fcfc021c58dc" />

peter-toth commented:
@cloud-fan, as you worked on `CollapseProject`, can you please take a look at this PR if you have some time?

peter-toth commented:
> Since this is a regression bug fix, we had better be clear in the PR title instead of `Improve CollapseProject`. Also, please revise the `Affected Versions` if you want to backport this to live release branches, @peter-toth .

Thanks @dongjoon-hyun. I renamed the PR and fixed the affected versions of the ticket.

dongjoon-hyun commented:
+1 for @cloud-fan 's opinion.

peter-toth commented:
Technically, I can create a new rule and limit the scope to batch eval expressions like Python UDFs, but I feel the new rule will share most of the logic with `CollapseProject`. Like checking if the producer is referenced at most once or if not then determining if it is cheap to inline the producer, as I don't want to move an expression that contains a Python UDF to the top project node in other cases.
I agree that the name `CollapseProject` doesn't fully describe the current functionality after this PR, but do we surely want create a new rule for a very similar logic?

cloud-fan commented:
@peter-toth can you elaborate on the cost model part? In `ExtractPythonUDFs`, we always deduplicate python UDFs, I think we can always merge Python UDFs into the top Project.

peter-toth commented:
@cloud-fan , @dongjoon-hyun , do you think we can proceed with this PR based on the above?

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1191,6 +1194,67 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
+    if (keep.isEmpty) {
+      (Seq.empty, substituted)
+    } else {
+      (substituted, keep ++ AttributeSet(substitute.flatMap(_.references)))
```
I do like this new approach that allows partial merging of two Projects. It's a net win and we should merge it first.

But we should also reconsider the cost model for python UDFs. Merging them into a single Project is a huge win, and we should do it even if we need to replicate some expensive expressions. We can discuss it in a followup PR.

peter-toth replied:
Ok, I can continue with a follow-up to merge all possible python UDFs after this PR.

peter-toth committed:
comment to `producerReferences` map

peter-toth commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:
```diff
@@ -1191,6 +1194,67 @@ object CollapseProject extends Rule[LogicalPlan] with AliasHelper {
+      producers: Seq[NamedExpression],
+      alwaysInline: Boolean): (Seq[NamedExpression], Seq[NamedExpression]) = {
+    lazy val producerAttributes = AttributeSet(producers.collect { case a: Alias => a.toAttribute })
+    lazy val producerReferences = AttributeMap(consumers
```
sure, added in: https://github.com/apache/spark/pull/52149/commits/0abacd79b4a5a2c135e3fb5df6c9ecc85761e78d

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/CollapseProjectSuite.scala:
```diff
@@ -298,4 +298,20 @@ class CollapseProjectSuite extends PlanTest {
+
+    val expected = testRelation
+      .select($"a" + 1 as "a_plus_1", $"b")
+      .select($"a_plus_1" + $"a_plus_1", $"b" + 1 as "b_plus_1")
```
Thinking about it again, if `b_plus_1` references 1000 columns, then this optimization actually makes the plan worse,  as the lower `Project` becomes super wide.

The benefits of this new optimization:
- if we happen to put two Python UDF into a single `Project`, it's a hude win.
- combining two expressions allows the optimizer to make better decisions.

Given that we will handle Python UDF specially later, let's be more conservative here:
```
...
val newRefs = AttributeSet(substitute.flatMap(_.references))
if (newRefs.length <= substitute) ... else (consumers, producers)
```

peter-toth replied:
1. Why actually does it cause problems if we have 1000 passthrough attributes in the lower project? And does it make sense to introduce a config (with a default like 200) to allow growing the lower project that wide?

2. I tihnk when we calculate the `newRefs` we should remove those passthrough attributes from the set that are already in `keep`. (This logic is missing from the current PR too.)

3. Also, if we have `b_plus_1` that references 1000 columns and `b_plus_2` that references only 1 column then we should keep only `b_plus_1` in the lower project, but substitute `b_plus_2` into the upper one. Actually I think we should probably sort the expressions in `substitute` by the net number of columns the expression's substitution would modify the lower project (this net number can be in the [-1..inf) range). Then we should retain the expressions in `substitute` until the running sum of those net values is <=0 and move the remaining expressions back to `keep`.

cloud-fan replied:
With whole stage codegen, a passthrough `Project` with many columns doesn't matter. But Spark already has many third-party native engines and each operator needs to materialize its output to columnar batches. Then such `Project` can hurt performance quite a bit.

cloud-fan replied:
Oh one more point: what if we already have two python UDFs in the lower project and we move one of them into the upper Project? Without merging the two `Project`, the benefit of combining two expressions is likely negligible. Moving expressions between two `Project`s should mainly consider Python UDFs.

peter-toth replied:
Yeah, indeed this is tough.
So shall we just force combining adjacent projects (when everything in them is deterministic) in a group of projects, if there is a parent project in the group with python UDFs?
